apps/els_lsp/src/els_refactorerl_utils.erl

```diff
@@ -0,0 +1,284 @@
+        {badrpc, _} ->
+          disable_node(Node),
+          error;
+        ok ->
+          receive
+            {ReqID, reply, {ok, _}} -> ok
+          end;
+        deny ->
+          notification("Adding is deined, retry!"), % TODO: Take this out overtime stability proves itself
+          timer:sleep(1000),
+          add(Uri, RecursionDepth + 1)
+      end;
+    _ ->
+      error
+  end;
+
+add(_Uri, RecursionDepth) when RecursionDepth >= ?MAX_RECURSION_DEPTH ->
+  notification("Cannot add module to RefactorErl!", ?MESSAGE_TYPE_ERROR),
+  error.
+
+  
+%%@doc
+%% Runs a Query on the RefactorErl node, returns its result if successfull
+%% If error happens, it just returns an empty list, as most of times the
+%% result then is directly converted to POIs.
+%% If the node timeouts or badrpc will come back, it disables the node and
+%% stills returns an empty list
+-spec query(string()) -> list().
+query(Query) ->
+  query(Query, 0).
+
+-spec query(string(), number()) -> list().
+query(Query, RecursionDepth) when RecursionDepth < ?MAX_RECURSION_DEPTH-> 
+  case referl_node() of 
+    {error, _} ->
+      [];
+    {ok, Node} ->
+      DisplayOpt = [{positions, linecol}, {output, msg}],
+      ReqID = request_id(),
+      Opts = [  self()
+              , ReqID
+              , {transform, semantic_query 
+                , [{ask_missing, false}
+                , {display_opt, DisplayOpt}
+                , {start_opt, []}
+                , {querystr, Query}]}
+             ],
+      Resp = rpc:call(Node, reflib_ui_router, request,  Opts),
+      case Resp of
+        {badrpc, _} ->
+          disable_node(Node),
+          []; % As most of times it is going to be processed right after.
+        ok ->
+          receive
+            {ReqID, reply, Result} -> Result
+          end;
+        deny ->
+          timer:sleep(1000),
+          query(Query, RecursionDepth + 1)
+      end
+  end;
+
+query(_Query, RecursionDepth) when RecursionDepth >= ?MAX_RECURSION_DEPTH ->
+  notification("Cannot execute query on RefactorErl!", ?MESSAGE_TYPE_ERROR),
+  [].
+
+
+%%@doc
+%% Util for popping up notifications
+-spec notification(string(), number()) -> atom().
+notification(Msg, Severity) ->
+  Param = #{ type => Severity,
+             message => list_to_binary(Msg) },
+  els_server:send_notification(<<"window/showMessage">>, Param).
+
+-spec notification(string()) -> atom().
+notification(Msg) ->
+  notification(Msg, ?MESSAGE_TYPE_INFO).
+
+
+%%@doc
+%% Creates a list of diagnostics form the RefactorErl results and a message
+%% Message can be the same, as this is called per diagnsotic type.
+%% The severity is only warning.
+-spec make_diagnostics(any(), string()) -> [els_diagnostics:diagnostic()].
+make_diagnostics([{{_Path, From, To}, Name} | Tail], DiagMsg) ->
+  Range = #{ from => From, to => To },
+  Id = refactorerl_poi,
+  #{ data := PoiData, range := PoiRange} =
+                        els_poi:new(Range, application, Id, Name), 
```

Comment:
I still think we shouldn't create a POI here. An application POI is supposed to be a function application (a call) and the ID is supposed to be the MFA or FA of the function. Here we are only interested in the range, which we already have, but in a slightly different form.

Reply:
Okay, removed the POI creation.